### PR TITLE
[fix] fix FSDP state_dict/load_state_dict for nested wrapped instances

### PR DESCRIPTION
**Pull request by myleott**
Similar to #434 but for FSDP now

### CODE REVIEW

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -384,18 +401,30 @@ def state_dict(self, *args, **kwargs):  # type: ignore
...
-            self._all_buffers_to(dtype=torch.float32)
-
-            state_dict = self.module.state_dict(*args, **kwargs)
-            # We copy the state_dict since full param will be freed after
-            # we exit the summon_full_params() context.
-            for key in state_dict.keys():
-                state_dict[key] = state_dict[key].clone()
+        # TODO
+        # Buffers dtype stays consistent with parameters.
+        # self._all_buffers_to(dtype=torch.float32)
```
Comment (myleott) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:406]: Removing these doesn't seem to cause any tests failures. It's not clear to me if this should be removed? Or if we should add more tests?

Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:406]: more likely I think we just don't have the test. I will file an issue.

Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:406]: or do you wanna address it here? I am fine either way.

Comment (myleott) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:406]: I see, we don't maintain the invariant that buffers are cast back to fp32 at the end of each `forward()`. Actually, I'm realizing we don't maintain fp32 buffers at all, so there could be loss of precision throughout training... we should fix that, perhaps in a separate PR: https://github.com/facebookresearch/fairscale/issues/442

Comment (sshleifer) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py]: self-assigned #442. I wrote the existing buffer logic with the incorrect assumption that buffers are static.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -172,11 +173,11 @@ def __init__(
         params = list(p for p in module.parameters() if not hasattr(p, "_is_sharded"))
 
         if self.flatten_parameters and len(params) > 0:
-            self.module: nn.Module = FlattenParamsWrapper(module, param_list=params)
+            self._fsdp_module: nn.Module = FlattenParamsWrapper(module, param_list=params)
```
Comment (sshleifer) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:176]: (nit) should this be called `_child_module/_wrapped_module/`? Does not seem guaranteed be an `FSDP` instance.

**Review by sshleifer — APPROVED** [review]
LGTM, nice tests!

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 2 above)
Comment (myleott) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:176]: yep, will rename to `_fsdp_wrapped_module`. Key is that it's easy to grep for in a state_dict, and unlikely to be in the user's module

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -235,6 +251,7 @@ def clip_grad_norm_(
         .. warning:: This needs to be called on all ranks, since synchronization
             primitives will be used.
         """
+        self._lazy_init()
```
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:254]: 👍

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -1011,3 +1072,23 @@ def alloc_storage_(data: torch.Tensor, size: torch.Size) -> None:
         return
     assert data.storage().size() == 0
     data.storage().resize_(size.numel())
+
+
+def _post_state_dict_hook(
+    module: nn.Module, state_dict: "OrderedDict[str, torch.Tensor]", prefix: str, *args: Any
+) -> "OrderedDict[str, torch.Tensor]":
+    if module.training_state == TrainingState.SUMMON_FULL_PARAMS:
+        # We copy the state_dict since full param will be freed after
```
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1081]: nice to put it in the hook!

**Review by min-xu-ai — APPROVED** [review]
awesome!
